semgrep-core/src/optimizing/Analyze_rule.ml

```diff
@@ -365,22 +365,22 @@ type step2 =
 type cnf_step2 = step2 cnf [@@deriving show]
 
 let or_step2 (Or xs) =
-  (* sanity check *)
-  xs
-  |> List.iter (function
-       | StringsAndMvars ([], _) -> raise GeneralPattern
-       | _ -> ());
-  let ys =
-    xs
-    |> Common.map (function
-         | StringsAndMvars (xs, _) -> Idents xs
-         | Regexp re -> Regexp2 re
-         | MvarRegexp (_mvar, re) -> Regexp2 re)
+  let step1_to_step2 =
+    List.map (function
+      | StringsAndMvars ([], _) -> raise GeneralPattern
+      | StringsAndMvars (xs, _) -> Idents xs
+      | Regexp re -> Regexp2 re
+      | MvarRegexp (_mvar, re) -> Regexp2 re)
   in
-  Or ys
+  (* Remove or cases where any of the possibilities is a general pattern *)
+  (* We need to do this because later, in the final regex generation,
+     empty cases will be disregarded *)
+  try Some (Or (step1_to_step2 xs)) with GeneralPattern -> None
 
 let and_step2 (And xs) =
-  let ys = xs |> Common.map or_step2 in
+  Common.pr2 (show_cnf_step1 (And xs));
+  let ys = xs |> List.filter_map or_step2 in
```

Comment:
Not crucial in this case you could add a changelog entry like "Improved filtering of rules based on file content".